### DISCUSSION TITLE
Including bespoke pre-processing steps

### DISCUSSION THREAD

**Discussion by OncologyDS** [jamesdolezal/slideflow#179] [Q&A] [answered]
Hi,

I've noticed that Slideflow allows for the easy creation of custom model architectures, but can't seem to find a similar functionality for bespoke pre-precessing steps: e.g. say I wanted to test a new blur detection method for tile selection at the extract tiles stage, I'm not sure how I could include this in the pipeline. Is there a way?

Thanks.

**Comment by jamesdolezal** [comment, marked as answer]
Thanks for your question. There are two places that pre-processing can be extended: during tile extraction (blur and artifact detection) and during training (augmentations).

**1. During tile extraction**: Slide-level filtering (such as blur and artifact detection) is performed with [`WSI.qc()`](https://github.com/jamesdolezal/slideflow/blob/6d72f06f1df2320b177ebab48f0d80de0b8656d2/slideflow/slide/__init__.py#L760). It is done by altering `WSI.grid`, a 2D array corresponding to locations where tiles will be extracted from the slide. Values of '1' indicate that the tile should be extracted at that location, and '0' indicates that the tile should *not* be extracted (e.g. because it was filtered out with blur detection or Otsu thresholding). If you want to implement a new slide-level processing step (such as a different blur detection method), you could extend the `WSI` class, adding an additional function for tile selection that modifies `WSI.grid`. You could then call `WSI.extract_tiles()` to extract tiles from an individual slide.

**2. During training**: We are working on an interface that will allow for customizing the image pre-processing steps (such as augmentations) during training. This interface is under development and planned for version 1.3 release.

I think that a tutorial explaining how to do (1) might be helpful, and I am happy to put one together. Could you make an issue with a feature request for a tutorial, so I can enter it into our queue?

**Reply by OncologyDS** [reply]
Thanks for the detailed reply. 

That sounds right to me, in which case I will pop in a feature request for a tutorial.

**Reply by OncologyDS** [reply]
As a follow-up thought: there are other pre-processing steps which I can think of. For example, I can imagine users wanting to only select tiles with a certain quantity of cell nuclei within them, or whose characteristics match those of a certain tissue. I imagine users might want to include steps such as these as well.
